Users system management features for Uberfire.

romartin opened the pull request:
- Not added the bpmspl ticket number into the commit message, is this message fine then?
- Included in this PR the users management testing webapp, it's a clean uberfire based webapp with the two users and groups management perspectives, if it has to be removed, np, just let me know. Thanks!

manstis commented:
I'll look properly tomorrow but my first comment is you need to add license headers...  correct headers based on recent BSIG emails.. sorry.

romartin commented:
Oh yes sorry! The key point is that i added the config for the headers.... hmmm, will check it now, thanks!

romartin commented:
Ok please review it when possible and I will fix all headers and more comments from the review if any, and submit a new single commit for it. Thanks!

romartin commented:
@psiroky Hey, checks have not passed.. but looking at the logs, the build is successful, any ideas?

romartin commented:
@psiroky oh sorry I just have seen the test failure... strange as I have not modified this module from the failed test, but will check it tomorrow. Thanks!

psiroky commented:
@romartin I believe this is an intermittent test failure. I've seen the test fail several times over the past few days (in different jobs). We need to take a look and fix the test. /cc @porcelli, @csadilek

csadilek commented:
@romartin I will start reviewing in my IDE (PR too big) so expect some spam/comments over the next hour or so :).

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-api/src/main/java/org/uberfire/ext/security/management/api/AbstractEntityManager.java:
```diff
@@ -0,0 +1,114 @@
+     * @return List of resulting entities from the search result. 
+     * @throws GenericUsersSystemManagementException
+     */
+    SearchResponse<T> search(SearchRequest request) throws GenericUsersSystemManagementException;
```
I think just SecurityManagamentException would suffice as name. The javadocs can point out that it's a general purpose exception.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-api/src/main/java/org/uberfire/ext/security/management/api/UsersManager.java:
```diff
@@ -0,0 +1,82 @@
+ * 
+ * @since 0.8.0
+ */
+public interface UsersManager extends AbstractEntityManager<User> {
```
Just a comment that I'd prefer singular UserManager instead of UsersManager but disregard if you prefer this.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-api/src/main/java/org/uberfire/ext/security/management/api/exception/EntityAlreadyExistsException.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @since 0.8.0
+ */
+@Portable
```
Since RuntimeException is serializable, we're getting warnings in all these sub types. Best to add @SuppressWarnings("serial"). We don't care about java.io serializability as we're using errai marshalling anyway.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-api/src/main/java/org/uberfire/ext/security/management/api/exception/GroupNotFoundException.java:
```diff
@@ -0,0 +1,21 @@
+@Portable
+public class GroupNotFoundException extends GenericUsersSystemManagementException {
+
+    private String identifier;
```
This field doesn't seem to be used.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-api/src/main/java/org/uberfire/ext/security/management/api/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,21 @@
+@Portable
+public class UserNotFoundException extends GenericUsersSystemManagementException {
+
+    private String identifier;
```
Also doesn't seem to be used.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-api/src/main/java/org/uberfire/ext/security/management/api/validation/RoleValidator.java:
```diff
@@ -0,0 +1,21 @@
+    
+    @Override
+    public Set<ConstraintViolation<Role>> validate(Role entity) {
+        return null;
```
Throw an exception to indicate that it's not yet implemented instead of returning null and causes failures later?

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/BackendUserSystemManager.java:
```diff
@@ -0,0 +1,161 @@
+                // Manager is not contextual.
+            }    
+        }
+        
```
Nice! We need to log an error though when the service is null and when an exception is thrown (casting) since these errors are non-recoverable.

romartin replied:
Will add a log when no user or group management service is available, but i think it's not necessary when the classcast exception occurs, as the service is available as well, but not initialized and destroyed in a given context.... but it should work as well

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/GroupsManagerServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public SearchResponse search(SearchRequest request) throws GenericUsersSystemManagementException {
```
Should be a SearchResponse<Group> (type parameter missing)

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/GroupsManagerServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+        return serviceImpl.getCapabilityStatus(capability);
+    }
+
+    private void handleGenericException(RuntimeException e) throws GenericUsersSystemManagementException {
```
We've already discusses this but for completeness. It's not clear what this method does. It shouldn't be needed (see below).

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/RolesManagerServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void handleGenericException(RuntimeException e) throws GenericUsersSystemManagementException {
+        LOG.error("Error executing roles management service.", e);
```
logic not needed.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/UsersManagerServiceImpl.java:
```diff
@@ -0,0 +1,136 @@
+            if (request.getPage() == 0) throw new IllegalArgumentException("First page must be 1.");
+            response = serviceImpl.search(request);
+        } catch (RuntimeException e) {
+            handleGenericException(e);
```
i.e. throw new SecurityManagementException(e);

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/UsersManagerServiceImpl.java:
```diff
@@ -0,0 +1,136 @@
+        return serviceImpl.getCapabilityStatus(capability);
+    }
+
+    private void handleGenericException(RuntimeException e) throws GenericUsersSystemManagementException {
```
logic not needed / also a duplicate.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/RolesManagerServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Override
+    public SearchResponse search(SearchRequest request) throws GenericUsersSystemManagementException {
```
type parameter missing. SearchResponse<Role>.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-backend/src/main/java/org/uberfire/ext/security/management/service/UsersManagerServiceImpl.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    @Override
+    public SearchResponse search(SearchRequest request)  {
```
type parameter: `SearchResponse<User>`

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-client/src/test/java/org/uberfire/ext/security/management/client/ClientUserSystemManagerTest.java:
```diff
@@ -0,0 +1,20 @@
+import org.mockito.runners.MockitoJUnitRunner;
+
+// TODO - @RunWith(MockitoJUnitRunner.class)
+@Ignore
```
Either delete this class or add tests?

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-client-wb/src/main/java/org/uberfire/ext/security/management/client/screens/home/UsersManagementHomeScreen.java:
```diff
@@ -0,0 +1,98 @@
+        view.show(UsersManagementWorkbenchConstants.INSTANCE.userEditorWelcomeText(),homeEnabledItems);
+    }
+
+    private void showError(final String message) {
```
method is not used.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-client-wb/src/main/java/org/uberfire/ext/security/management/client/screens/home/GroupsManagementHomeScreen.java:
```diff
@@ -0,0 +1,99 @@
+        view.show(UsersManagementWorkbenchConstants.INSTANCE.groupEditorWelcomeText(),homeEnabledItems);
+    }
+
+    private void showError(final String message) {
```
method is not used.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-keycloak/src/main/java/org/uberfire/ext/security/management/keycloak/KeyCloakGroupsManager.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Override
+    public SearchResponse search(SearchRequest request) throws GenericUsersSystemManagementException {
```
type parameter of return type missing.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-keycloak/src/main/java/org/uberfire/ext/security/management/keycloak/KeyCloakUsersManager.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Override
+    public SearchResponse search(SearchRequest request) throws GenericUsersSystemManagementException {
```
return type type param missing.

csadilek commented:
@romartin I think this is really good work! Couple of minor issues in comments inline (see above). There's good test coverage on the various SPI implementations, but I think the backend project could need some more tests.

As @manstis pointed out, the license headers are missing and where present they're incorrect (copied from other project and therefore have the wrong year etc.).

Many files have unused imports. Please execute an organize imports refactoring on the projects to get rid of those warnings.

For the showcase, I think we should keep it until security management is integrated into the drools and jbpm workbench.

Can you update the commit with these minor changes and maybe some more tests for the backend? I think it's good to go then. Thanks.

csadilek commented on uberfire-security/uberfire-security-management/uberfire-security-management-keycloak/src/main/java/org/uberfire/ext/security/management/keycloak/KeyCloakGroupsManager.java:
```diff
@@ -0,0 +1,155 @@
+                case CAN_ADD_GROUP:
+                case CAN_READ_GROUP:
+                case CAN_DELETE_GROUP:
+                    return CapabilityStatus.ENABLED;
```
I think adding a default case here is cleaner and gets rid of the warning. case default: return CapabilityStatus.UNSUPPORTED or simply break.

romartin commented:
Added a new commit that fixes some of the above minor issues:
    - property file for specifying a default provider in the assemblies
    - usersManager, groupsManager, rolesManager and all impls renamed as singular "userManager", "groupManager", "roleManager", etc
    - Renamed GenericUserSystemManagement exception to SecurityManagementException
    - unused imports clean up
    - other minor issues

romartin commented:
Note that I will add some more commits and will squash all them when finished into a single one :)

romartin commented:
Added another commits that adds all the missing license headers.

romartin commented:
Added another commit that refactor how services are specified. Instead of specifying users and roles service impls using different properties, just use a single property that points to a bean that provider the different services for the concrete provider impl.

romartin commented:
Tomorrow i will add more backend tests, squash the commits and the PR should be fine to merge.  Will keep you up to date. Thanks!

csadilek commented:
Thanks @romartin, changes all look good!

romartin commented:
Ok @csadilek, I think I have fixed all issues:

- Added license headers
- Added property file for specifying a default provider in the assemblies
-  Refactor manager names
- Other refactors and issues
- Added tests for all missing methods for the backend different providers
- Imports clean-up
- Fixed ClientUserSystemManagerTest as it was empty/ignored

Thanks!

csadilek commented:
+1

csadilek commented:
Thanks @romartin. 

master: ee493acacf9e24e79368556c9498b5743d90b66f

manstis commented:
@romartin Could you please add ```.gitignore``` files before someone accidentally commits ```/target``` or ```WEB-INF/classes/``` folders later on :)

romartin commented:
Ohh yess @manstis I forget them sorry, added in this PR -> https://github.com/uberfire/uberfire-extensions/pull/153 
Thanks! :)
